### PR TITLE
fix: enable torchembedding deepcopy

### PR DESCRIPTION
**Pull request by AnneYang720**
Goals:

This PR is related to #1131

### CODE REVIEW

**Comment by AnneYang720** [comment]
After the modification
```python
import torch
from docarray import BaseDocument
from docarray.typing.tensor.embedding.torch import TorchEmbedding

class Meow(BaseDocument):
    embedding : TorchEmbedding

a = Meow(embedding=torch.randn(42))
print(id(a.embedding), id(a.copy(deep=True).embedding))
```
output has no error
```
140248435632272 140248435636032
```

**Comment by samsja** [comment]
Can u add to test to check that it is actual deep copied and not shallow copy ?

File: tests/units/typing/tensor/test_torch_tensor.py [hunk 1]
```diff
@@ -152,3 +152,17 @@ def test_parametrized_operations():
...
+    from docarray import BaseDocument
+
+    class MMdoc(BaseDocument):
+        embedding: TorchEmbedding
+
+    doc = MMdoc(embedding=torch.randn(32))
+    doc_copy = doc.copy(deep=True)
+
+    assert id(doc) != id(doc_copy)
+    assert id(doc.embedding) != id(doc_copy.embedding)
```
Comment (samsja) [inline tests/units/typing/tensor/test_torch_tensor.py:167]: I think that this test is not enough.

Two pytorch variable can have different ID, which mean they are different variable, but still point to the same underlying data storage. 

You can use `tensor.data_ptr()` to get the pytorch pointer to the actual data (we use it in a couple of otehr test).


Another thing that you can do is to actually check that if you changed the value of `doc_copy` it does not affect `doc`

**Comment by JohannesMessner** [comment]
Can you explain this fix briefly? It is not obvious to me why this is fixing it

**Comment by AnneYang720** [comment]
> Can you explain this fix briefly? It is not obvious to me why this is fixing it

The deepcopy error is thrown from here https://github.com/pytorch/pytorch/blob/master/torch/_tensor.py#L170, and line above is 
```python
new_tensor = self.new_empty([])
if type(new_tensor) is not type(self):
```
where `type(new_tensor)` is  `<class 'torch.Tensor'>` and 
`type(self)` is `<class 'docarray.typing.tensor.embedding.torch.TorchEmbedding'>`.

So according to the error msg, I should implement `new_empty()` for the subclass, which means the `new_tensor` should also be a `TorchEmbedding`.

**Comment by samsja** [comment]
> > Can you explain this fix briefly? It is not obvious to me why this is fixing it
> 
> The deepcopy error is thrown from here https://github.com/pytorch/pytorch/blob/master/torch/_tensor.py#L170, and lines above are
> 
> ```python
> new_tensor = self.new_empty([])
> if type(new_tensor) is not type(self):
> ```
> 
> where `type(new_tensor)` is `<class 'torch.Tensor'>` and `type(self)` is `<class 'docarray.typing.tensor.embedding.torch.TorchEmbedding'>`.
> 
> So according to the error msg, I should implement `new_empty()` for the subclass, which means the `new_tensor` should also be a `TorchEmbedding`.

@AnneYang720  can u add this to the docstring of `new_empty` in your PR so that ppl who will read the code will understand the same ?

**Review by samsja — APPROVED** [review]

**Pull request merged**